File: DataFeedViews/InitialConditionsNonSpatialDataFeedView.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Core
Imports System.Globalization
Imports SyncroSim.Core.Forms
Imports System.Windows.Forms
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class InitialConditionsNonSpatialDataFeedView

    Protected Overrides Sub InitializeView()

        MyBase.InitializeView()

        Dim v As DataFeedView = Me.Session.CreateMultiRowDataFeedView(
            Me.Scenario, Me.ControllingScenario)

        Me.PanelDistribution.Controls.Add(v)

    End Sub

    Public Overrides Sub LoadDataFeed(ByVal dataFeed As DataFeed)

        MyBase.LoadDataFeed(dataFeed)

        Me.SetTextBoxBinding(Me.TextBoxTotalAmount, DATASHEET_NSIC_TOTAL_AMOUNT_COLUMN_NAME)
        Me.SetTextBoxBinding(Me.TextBoxNumCells, DATASHEET_NSIC_NUM_CELLS_COLUMN_NAME)
        Me.SetCheckBoxBinding(Me.CheckBoxCalcFromDist, DATASHEET_NSIC_CALC_FROM_DIST_COLUMN_NAME)

        Dim v As DataFeedView = CType(Me.PanelDistribution.Controls(0), DataFeedView)
        v.LoadDataFeed(dataFeed, DATASHEET_NSIC_DISTRIBUTION_NAME)

        Me.RefreshBoundControls()

        If (Me.CheckBoxCalcFromDist.Checked) Then
            Me.SetNumCellsFromDistribution()
        End If

        Me.CalculateCellSize()

    End Sub

    Protected Overrides Sub OnDataSheetChanged(e As DataSheetMonitorEventArgs)

        MyBase.OnDataSheetChanged(e)

        Dim amountlabel As String = Nothing
        Dim units As TerminologyUnit
        Dim unitsLbl As String = Nothing

        GetAmountLabelTerminology(e.DataSheet, amountlabel, units)
        unitsLbl = TerminologyUnitToString(units).ToLower(CultureInfo.CurrentCulture)

        Me.LabelTotalAmount.Text = String.Format(CultureInfo.CurrentCulture, "Total ({0}):", unitsLbl)
        Me.LabelCellSize.Text = String.Format(CultureInfo.CurrentCulture, "Cell size ({0}):", unitsLbl)
        Me.TextBoxNumCells.Enabled = (Me.ShouldEnableView And (Not Me.CheckBoxCalcFromDist.Checked))

    End Sub

    Public Overrides Sub EnableView(enable As Boolean)

        If (Me.PanelDistribution.Controls.Count > 0) Then

            Dim v As DataFeedView = CType(Me.PanelDistribution.Controls(0), DataFeedView)
            v.EnableView(enable)

        End If

        Me.TextBoxTotalAmount.Enabled = enable
        Me.TextBoxNumCells.Enabled = (enable And (Not Me.CheckBoxCalcFromDist.Checked))
        Me.CheckBoxCalcFromDist.Enabled = enable
        Me.ButtonClearAll.Enabled = enable

    End Sub

    Protected Overrides Sub OnRowsAdded(sender As Object, e As Core.DataSheetRowEventArgs)

        MyBase.OnRowsAdded(sender, e)
        Me.RecomputeNumCellsForDistributionChange(sender)

    End Sub

    Protected Overrides Sub OnRowsDeleted(sender As Object, e As Core.DataSheetRowEventArgs)

        MyBase.OnRowsDeleted(sender, e)
        Me.RecomputeNumCellsForDistributionChange(sender)

    End Sub

    Protected Overrides Sub OnRowsModified(sender As Object, e As Core.DataSheetRowEventArgs)

        MyBase.OnRowsModified(sender, e)
        Me.RecomputeNumCellsForDistributionChange(sender)

    End Sub

    Protected Overrides Sub OnBoundTextBoxChanged(textBox As TextBox, columnName As String)

        MyBase.OnBoundTextBoxChanged(textBox, columnName)
        Me.CalculateCellSize()

    End Sub

    Protected Overrides Sub OnBoundCheckBoxChanged(checkBox As CheckBox, columnName As String)

        MyBase.OnBoundCheckBoxChanged(checkBox, columnName)

        If (Me.CheckBoxCalcFromDist.Checked) Then

            Me.TextBoxNumCells.Enabled = False
            Me.SetNumCellsFromDistribution()

        Else

            Me.TextBoxNumCells.Enabled = True
            Me.SetTextBoxData(Me.TextBoxNumCells, Nothing)

        End If

        Me.CalculateCellSize()

    End Sub

    Private Sub RecomputeNumCellsForDistributionChange(ByVal sender As Object)

        Dim ds As DataSheet = Me.DataFeed.GetDataSheet(DATASHEET_NSIC_DISTRIBUTION_NAME)

        If (sender Is ds And Me.Scenario Is Me.ControllingScenario) Then

            If (Me.CheckBoxCalcFromDist.Checked) Then

                Me.SetNumCellsFromDistribution()
                Me.CalculateCellSize()

            End If

        End If

    End Sub

    Private Sub SetNumCellsFromDistribution()

        Debug.Assert(Me.CheckBoxCalcFromDist.Checked)

        Dim NumCells As Integer = Me.CalculateNumCellsFromDistribution()

        If (NumCells > 0) Then
            Me.SetTextBoxData(Me.TextBoxNumCells, NumCells.ToString("N4", CultureInfo.CurrentCulture))
        Else
            Me.SetTextBoxData(Me.TextBoxNumCells, Nothing)
        End If

    End Sub

    Private Function CalculateNumCellsFromDistribution() As Integer

        Dim NumCells As Integer = 0
        Dim dt As DataTable = Me.DataFeed.GetDataSheet(DATASHEET_NSIC_DISTRIBUTION_NAME).GetData()

        ' Use just the lowest(1st iteration) entered to deal with multiple iterations
        Dim minIteration As Integer = 0
        For Each dr As DataRow In dt.Rows

            If (dr.RowState <> DataRowState.Deleted) Then
                If IsDBNull(dr(DATASHEET_ITERATION_COLUMN_NAME)) Then
                    minIteration = 0
                Else
                    minIteration = Math.Min(minIteration, CInt(dr(DATASHEET_ITERATION_COLUMN_NAME)))
                End If

            End If

        Next

        For Each dr As DataRow In dt.Rows

            If (dr.RowState <> DataRowState.Deleted) Then

                Dim iteration = CInt(IIf(IsDBNull(dr(DATASHEET_ITERATION_COLUMN_NAME)), 0, dr(DATASHEET_ITERATION_COLUMN_NAME)))
                If iteration = minIteration Then
                    Dim val As Double = CDbl(dr(DATASHEET_NSIC_DISTRIBUTION_RELATIVE_AMOUNT_COLUMN_NAME))
                    NumCells += CInt(Math.Round(val))
                End If

            End If

        Next

        Return NumCells

    End Function

    Private Sub CalculateCellSize()

        Dim ns As String = Me.TextBoxNumCells.Text.Trim()
        Dim ta As String = Me.TextBoxTotalAmount.Text.Trim()

        If (String.IsNullOrEmpty(ns) Or String.IsNullOrEmpty(ta)) Then
            Me.TextBoxCellSize.Text = Nothing
            Return
        End If

        Dim TotalAmount As Double = 0.0
        Dim NumCells As Integer = 0

        If (Not Double.TryParse(ta, NumberStyles.Any, CultureInfo.InvariantCulture, TotalAmount)) Then
            Me.TextBoxCellSize.Text = Nothing
            Return
        End If

        If (Not Integer.TryParse(ns, NumberStyles.Any, CultureInfo.InvariantCulture, NumCells)) Then
            Me.TextBoxCellSize.Text = Nothing
            Return
        End If

        Dim CellSize As Double

        If (TotalAmount = 0.0 Or NumCells = 0.0) Then
            CellSize = 0.0
        Else
            CellSize = TotalAmount / NumCells
        End If

        Me.TextBoxCellSize.Text = CellSize.ToString("N4", CultureInfo.CurrentCulture)

    End Sub

    Private Sub ButtonClearAll_Click(sender As Object, e As EventArgs) Handles ButtonClearAll.Click

        Me.ResetBoundControls()
        Me.DataFeed.DataSheets(DATASHEET_NSIC_NAME).ClearData()
        Me.TextBoxCellSize.Text = Nothing

    End Sub

End Class

File: src/DataFeedViews/InitialConditionsNonSpatialDataFeedView.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2024 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using System.Diagnostics;
using System.Globalization;
using System.Windows.Forms;
using SyncroSim.Core;
using SyncroSim.Core.Forms;

namespace SyncroSim.STSim
{
    internal partial class InitialConditionsNonSpatialDataFeedView
    {
        public InitialConditionsNonSpatialDataFeedView()
        {
            InitializeComponent();
        }

        protected override void InitializeView()
        {
            base.InitializeView();

            DataFeedView v = this.Session.CreateMultiRowDataFeedView(this.Scenario, this.ControllingScenario);
            this.PanelDistribution.Controls.Add(v);
        }

        public override void LoadDataFeed(DataFeed dataFeed)
        {
            base.LoadDataFeed(dataFeed);

            this.SetTextBoxBinding(this.TextBoxTotalAmount, Strings.DATASHEET_NSIC_TOTAL_AMOUNT_COLUMN_NAME);
            this.SetTextBoxBinding(this.TextBoxNumCells, Strings.DATASHEET_NSIC_NUM_CELLS_COLUMN_NAME);
            this.SetCheckBoxBinding(this.CheckBoxCalcFromDist, Strings.DATASHEET_NSIC_CALC_FROM_DIST_COLUMN_NAME);

            DataFeedView v = (DataFeedView)this.PanelDistribution.Controls[0];
            v.LoadDataFeed(dataFeed, Strings.DATASHEET_NSIC_DISTRIBUTION_NAME);

            this.RefreshBoundControls();
            this.CalculateCellSize();
        }

        protected override void OnDataSheetChanged(DataSheetMonitorEventArgs e)
        {
            base.OnDataSheetChanged(e);

            string amountlabel = null;
            TerminologyUnit units = 0;
            string unitsLbl = null;

            TerminologyUtilities.GetAmountLabelTerminology(this.Project, ref amountlabel, ref units);
            unitsLbl = TerminologyUtilities.TerminologyUnitToString(units).ToLower(CultureInfo.InvariantCulture);

            this.LabelTotalAmount.Text = string.Format(CultureInfo.InvariantCulture, "Total ({0}):", unitsLbl);
            this.LabelCellSize.Text = string.Format(CultureInfo.InvariantCulture, "Cell size ({0}):", unitsLbl);
            this.TextBoxNumCells.Enabled = (this.ShouldEnableView() && (!this.CheckBoxCalcFromDist.Checked));
        }

        public override void EnableView(bool enable)
        {
            if (this.PanelDistribution.Controls.Count > 0)
            {
                DataFeedView v = (DataFeedView)this.PanelDistribution.Controls[0];
                v.EnableView(enable);
            }

            this.TextBoxTotalAmount.Enabled = enable;
            this.TextBoxNumCells.Enabled = (enable && (!this.CheckBoxCalcFromDist.Checked));
            this.CheckBoxCalcFromDist.Enabled = enable;
            this.ButtonClearAll.Enabled = enable;
        }

        protected override void OnRowsAdded(object sender, Core.DataSheetRowEventArgs e)
        {
            base.OnRowsAdded(sender, e);
            this.RecomputeNumCellsForDistributionChange(sender);
        }

        protected override void OnRowsDeleted(object sender, Core.DataSheetRowEventArgs e)
        {
            base.OnRowsDeleted(sender, e);
            this.RecomputeNumCellsForDistributionChange(sender);
        }

        protected override void OnRowsModified(object sender, Core.DataSheetRowEventArgs e)
        {
            base.OnRowsModified(sender, e);
            this.RecomputeNumCellsForDistributionChange(sender);
        }

        protected override void OnBoundTextBoxChanged(TextBox textBox, string columnName)
        {
            base.OnBoundTextBoxChanged(textBox, columnName);
            this.CalculateCellSize();
        }

        protected override void OnBoundCheckBoxChanged(CheckBox checkBox, string columnName)
        {
            base.OnBoundCheckBoxChanged(checkBox, columnName);

            if (this.CheckBoxCalcFromDist.Checked)
            {
                this.TextBoxNumCells.Enabled = false;
                this.SetNumCellsFromDistribution();
            }
            else
            {
                this.TextBoxNumCells.Enabled = true;
                this.SetTextBoxData(this.TextBoxNumCells, null);
            }

            this.CalculateCellSize();
        }

        private void RecomputeNumCellsForDistributionChange(object sender)
        {
            DataSheet ds = this.DataFeed.GetDataSheet(Strings.DATASHEET_NSIC_DISTRIBUTION_NAME);

            if (sender == ds && this.Scenario == this.ControllingScenario)
            {
                if (this.CheckBoxCalcFromDist.Checked)
                {
                    this.SetNumCellsFromDistribution();
                    this.CalculateCellSize();
                }
            }
        }

        private void SetNumCellsFromDistribution()
        {
            Debug.Assert(this.CheckBoxCalcFromDist.Checked);
            int NumCells = this.CalculateNumCellsFromDistribution();

            if (NumCells > 0)
            {
                this.SetTextBoxData(this.TextBoxNumCells, NumCells.ToString("N4", CultureInfo.InvariantCulture));
            }
            else
            {
                this.SetTextBoxData(this.TextBoxNumCells, null);
            }
        }

        private int CalculateNumCellsFromDistribution()
        {
            int NumCells = 0;
            DataTable dt = this.DataFeed.GetDataSheet(Strings.DATASHEET_NSIC_DISTRIBUTION_NAME).GetData();

            // Use just the lowest(1st iteration) entered to deal with multiple iterations
            int minIteration = 0;
            foreach (DataRow dr in dt.Rows)
            {
                if (dr.RowState != DataRowState.Deleted)
                {
                    if (Convert.IsDBNull(dr[Strings.DATASHEET_ITERATION_COLUMN_NAME]))
                    {
                        minIteration = 0;
                    }
                    else
                    {
                        minIteration = Math.Min(
                            minIteration, 
                            Convert.ToInt32(dr[Strings.DATASHEET_ITERATION_COLUMN_NAME], CultureInfo.InvariantCulture));
                    }
                }
            }

            foreach (DataRow dr in dt.Rows)
            {
                if (dr.RowState != DataRowState.Deleted)
                {
                    var iteration = Convert.ToInt32(
                        Convert.IsDBNull(dr[Strings.DATASHEET_ITERATION_COLUMN_NAME]) ? 0 : 
                        dr[Strings.DATASHEET_ITERATION_COLUMN_NAME], CultureInfo.InvariantCulture);

                    if (iteration == minIteration)
                    {
                        double val = Convert.ToDouble(
                            dr[Strings.DATASHEET_NSIC_DISTRIBUTION_RELATIVE_AMOUNT_COLUMN_NAME], 
                            CultureInfo.InvariantCulture);

                        NumCells += Convert.ToInt32(Math.Round(val), CultureInfo.InvariantCulture);
                    }
                }
            }

            return NumCells;
        }

        private void CalculateCellSize()
        {
            string ns = this.TextBoxNumCells.Text.Trim();
            string ta = this.TextBoxTotalAmount.Text.Trim();

            if (string.IsNullOrEmpty(ns) | string.IsNullOrEmpty(ta))
            {
                this.TextBoxCellSize.Text = null;
                return;
            }

            double TotalAmount = 0.0;
            int NumCells = 0;

            if (!double.TryParse(ta, NumberStyles.Any, CultureInfo.InvariantCulture, out TotalAmount))
            {
                this.TextBoxCellSize.Text = null;
                return;
            }

            if (!int.TryParse(ns, NumberStyles.Any, CultureInfo.InvariantCulture, out NumCells))
            {
                this.TextBoxCellSize.Text = null;
                return;
            }

            double CellSize = 0;

            if (TotalAmount == 0.0 || NumCells == 0.0)
            {
                CellSize = 0.0;
            }
            else
            {
                CellSize = TotalAmount / NumCells;
            }

            this.TextBoxCellSize.Text = CellSize.ToString("N4", CultureInfo.InvariantCulture);
        }

        private void ButtonClearAll_Click(object sender, EventArgs e)
        {
            this.ResetBoundControls();
            this.DataFeed.DataSheets[Strings.DATASHEET_NSIC_NAME].ClearData();
            this.TextBoxCellSize.Text = null;
        }
    }
}
